components/ui/pathogen-map/pathogen-map-popup.tsx

```diff
@@ -0,0 +1,102 @@
+import { MapRef, Popup, useMap } from "react-map-gl";
+
+import { PathogenDataPointPropertiesBase } from "./pathogen-map";
+
+export type PopupContentGenerator<
+  TPathogenDataPointProperties extends PathogenDataPointPropertiesBase
+> = (input: TPathogenDataPointProperties) => React.ReactNode;
+
+type VisiblePopupInfo<
+  TPathogenDataPointProperties extends PathogenDataPointPropertiesBase
+> = { visible: true; properties: TPathogenDataPointProperties };
+type HiddenPopupInfo<
+  TPathogenDataPointProperties extends PathogenDataPointPropertiesBase
+> = { visible: false; properties: null };
+
+export type PopupInfo<
+  TPathogenDataPointProperties extends PathogenDataPointPropertiesBase
+> =
+  | VisiblePopupInfo<TPathogenDataPointProperties>
+  | HiddenPopupInfo<TPathogenDataPointProperties>;
+
+interface PathogenMapPopupProps<
+  TPathogenDataPointProperties extends PathogenDataPointPropertiesBase
+> {
+  mapId: string;
+  popUpInfo: PopupInfo<TPathogenDataPointProperties>;
+  generatePopupContent: PopupContentGenerator<TPathogenDataPointProperties>;
+}
+
+export function PathogenMapPopup<
+  TPathogenDataPointProperties extends PathogenDataPointPropertiesBase
+>({
+  mapId,
+  popUpInfo,
+  generatePopupContent,
+}: PathogenMapPopupProps<TPathogenDataPointProperties>) {
+  const allMaps = useMap();
+  const map = allMaps[mapId];
+
+  const getMapboxLatitudeOffset = (map: MapRef) => {
+    // Map seems to zoom in in powers of 2, so reducing offset by powers of 2 keeps the modal apprximately
+    // in the same center everytime
+    // offset needs to reduce exponentially with zoom -- higher zoom x smaller offset
+    let mapZoom = map.getZoom();
+
+    return 80 / Math.pow(2, mapZoom);
+  };
+
+  if (popUpInfo.visible === false) {
+    return null;
+  }
+
+  // Not a typo, there is some kind of underlying problem here worth investigating that
+  // makes switching the latitude and longitude a required step to produce the correct results.
+  const latitude = popUpInfo.properties.longitude ?? 0;
+  const longitude = popUpInfo.properties.latitude ?? 0;
```

Comment:
Highlighting one of those "questionable things" I mentioned at the start.
I don't know why this gives the correct results but it does.
I assume there's some kind of underlying problem here that needs investigation, I'd be happy to take this away and see why exactly it's like this and how we can clean it up. I'm hesitant about doing it here because I don't know how much of a rabbit hole it is so I was planning on leaving this as-is for this PR if that's alright with everyone.